File: lib/php/tests/MetricsTest.php
<?php

declare(strict_types=1);

namespace WgPortal\Client\Tests;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use PHPUnit\Framework\TestCase;
use WgPortal\Client\HttpClient;
use WgPortal\Client\MetricsApi;

class MetricsTest extends TestCase
{
    private const BASE_URL = 'http://wg-portal.test:8888';
    private const PEER_ID = 'xTIBA5rboUvnH4htodjb6e697QjLERt1NAB4mZqp8Dg=';

    private function makeApi(array &$queue): MetricsApi
    {
        $ref = new \ReflectionClass(HttpClient::class);
        $http = $ref->newInstanceWithoutConstructor();

        $prop = $ref->getProperty('baseUrl');
        $prop->setAccessible(true);
        $prop->setValue($http, self::BASE_URL . '/api/v1');

        $headersProp = $ref->getProperty('headers');
        $headersProp->setAccessible(true);
        $headersProp->setValue($http, ['Authorization' => 'Basic dGVzdDp0ZXN0']);

        $mock = new MockHandler($queue);
        $httpProp = $ref->getProperty('http');
        $httpProp->setAccessible(true);
        $httpProp->setValue($http, new GuzzleClient(['handler' => HandlerStack::create($mock)]));

        return new MetricsApi($http);
    }

    public function testByInterface(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);
        $queue[] = new Response(200, ['Content-Type' => 'application/json'], json_encode([
            'InterfaceIdentifier' => 'wg0',
            'BytesReceived' => 123456789,
            'BytesTransmitted' => 987654321,
        ]));

        $result = $api->byInterface('wg0');
        $this->assertEquals('wg0', $result['InterfaceIdentifier']);
        $this->assertEquals(123456789, $result['BytesReceived']);
    }

    public function testByPeer(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);
        $queue[] = new Response(200, ['Content-Type' => 'application/json'], json_encode([
            'PeerIdentifier' => self::PEER_ID,
            'BytesReceived' => 1000,
            'BytesTransmitted' => 2000,
            'IsPingable' => true,
        ]));

        $result = $api->byPeer(self::PEER_ID);
        $this->assertEquals(self::PEER_ID, $result['PeerIdentifier']);
        $this->assertTrue($result['IsPingable']);
    }

    public function testByUser(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);
        $queue[] = new Response(200, ['Content-Type' => 'application/json'], json_encode([
            'UserIdentifier' => 'uid-1234567',
            'BytesReceived' => 11000,
            'BytesTransmitted' => 22000,
            'PeerCount' => 2,
            'PeerMetrics' => [],
        ]));

        $result = $api->byUser('uid-1234567');
        $this->assertEquals('uid-1234567', $result['UserIdentifier']);
        $this->assertEquals(2, $result['PeerCount']);
    }
}

File: lib/rust/tests/metrics_test.rs
use mockito::Server;
use serde_json::json;
use wg_portal_client::WgPortalClient;

fn setup() -> (mockito::ServerGuard, WgPortalClient) {
    let server = Server::new();
    let client = WgPortalClient::new(&server.url(), "admin@wgportal.local", "test-token");
    (server, client)
}

const PEER_ID: &str = "xTIBA5rboUvnH4htodjb6e697QjLERt1NAB4mZqp8Dg=";

#[test]
fn by_interface() {
    let (mut server, client) = setup();
    let mock = server
        .mock("GET", "/api/v1/metrics/by-interface/wg0")
        .with_body(
            json!({
                "InterfaceIdentifier": "wg0",
                "BytesReceived": 123456789,
                "BytesTransmitted": 987654321
            })
            .to_string(),
        )
        .with_header("content-type", "application/json")
        .create();

    let result = client.metrics().by_interface("wg0").unwrap();
    assert_eq!(result.interface_identifier, "wg0");
    assert_eq!(result.bytes_received, 123456789);
    mock.assert();
}

#[test]
fn by_peer() {
    let (mut server, client) = setup();
    let mock = server
        .mock("GET", format!("/api/v1/metrics/by-peer/{}", PEER_ID).as_str())
        .with_body(
            json!({
                "PeerIdentifier": PEER_ID,
                "BytesReceived": 1000,
                "BytesTransmitted": 2000,
                "Endpoint": "12.34.56.78",
                "IsPingable": true
            })
            .to_string(),
        )
        .with_header("content-type", "application/json")
        .create();

    let result = client.metrics().by_peer(PEER_ID).unwrap();
    assert_eq!(result.peer_identifier, PEER_ID);
    assert_eq!(result.is_pingable, Some(true));
    mock.assert();
}

#[test]
fn by_user() {
    let (mut server, client) = setup();
    let mock = server
        .mock("GET", "/api/v1/metrics/by-user/uid-1234567")
        .with_body(
            json!({
                "UserIdentifier": "uid-1234567",
                "BytesReceived": 11000,
                "BytesTransmitted": 22000,
                "PeerCount": 2,
                "PeerMetrics": [{
                    "PeerIdentifier": PEER_ID,
                    "BytesReceived": 1000,
                    "BytesTransmitted": 2000
                }]
            })
            .to_string(),
        )
        .with_header("content-type", "application/json")
        .create();

    let result = client.metrics().by_user("uid-1234567").unwrap();
    assert_eq!(result.user_identifier, "uid-1234567");
    assert_eq!(result.peer_count, 2);
    assert_eq!(result.peer_metrics.len(), 1);
    mock.assert();
}
